File: app/src/main/java/com/system/m4/businness/SmsBroadcastReceiver.java
package com.system.m4.businness;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.telephony.SmsMessage;

import com.system.m4.kotlin.services.SMSReaderBusiness;
import com.system.m4.kotlin.transaction.TransactionBusiness;
import com.system.m4.views.vos.Transaction;

public class SmsBroadcastReceiver extends BroadcastReceiver {

    public static final String SMS_BUNDLE = "pdus";

    public void onReceive(Context context, Intent intent) {

        Bundle intentExtras = intent.getExtras();

        if (intentExtras != null) {

            Object[] sms = (Object[]) intentExtras.get(SMS_BUNDLE);

            if (sms != null) {

                for (Object sm : sms) {

                    String format = intentExtras.getString("format");
                    SmsMessage smsMessage = SmsMessage.createFromPdu((byte[]) sm, format);

                    Transaction transaction = SMSReaderBusiness.Companion.readSMS(context, smsMessage.getMessageBody());
                    if (transaction != null) {
                        new TransactionBusiness().save(transaction, null);
                    }
                }
            }
        }
    }
}

File: app/src/main/java/com/system/m4/kotlin/services/SmsBroadcastReceiver.kt
package com.system.m4.kotlin.services

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.telephony.SmsMessage
import com.system.m4.kotlin.transaction.TransactionBusiness

class SmsBroadcastReceiver : BroadcastReceiver() {

    override fun onReceive(context: Context, intent: Intent) {

        val intentExtras = intent.extras

        if (intentExtras != null) {

            val sms = intentExtras.get(SMS_BUNDLE) as Array<*>

            for (sm in sms) {

                val format = intentExtras.getString(SMS_FORMAT)
                val smsMessage = SmsMessage.createFromPdu(sm as ByteArray, format)

                val transaction = SmsReaderBusiness.readSMS(context, smsMessage.messageBody)
                if (transaction != null) {
                    TransactionBusiness.save(context, transaction, null)
                }
            }
        }
    }

    companion object {

        val SMS_BUNDLE = "pdus"
        val SMS_FORMAT = "format"
    }
}